File: bench/conc.c
/* Concurrency benchmark (C, pthreads) — the fork-join baseline: 4 workers each
 * compute fib(N), the parent sums. `./conc N T`: T<=1 runs the four sequentially
 * (the 1-thread baseline), else one pthread per worker. Compare against the same
 * workload in Rust (bench/conc.rs) and Axion (bench/conc.axi, session tasks on
 * the M:N scheduler). Naive fib so the compute dominates. */
#include <pthread.h>
#include <stdio.h>
#include <stdlib.h>

/* volatile so the optimizer cannot hoist/CSE the four `fib(ARG)` calls in the
 * sequential path (they must each really run, like the four session workers). */
static volatile long ARG;
static long fib(long n) { return n < 2 ? n : fib(n - 1) + fib(n - 2); }
static void *worker(void *r) {
  *(long *)r = fib(ARG);
  return NULL;
}

int main(int argc, char **argv) {
  ARG = argc > 1 ? atol(argv[1]) : 34;
  int t = argc > 2 ? atoi(argv[2]) : 4;
  long res[4];
  if (t <= 1) {
    for (int i = 0; i < 4; i++) res[i] = fib(ARG);
  } else {
    pthread_t th[4];
    for (int i = 0; i < 4; i++) pthread_create(&th[i], NULL, worker, &res[i]);
    for (int i = 0; i < 4; i++) pthread_join(th[i], NULL);
  }
  printf("%ld\n", res[0] + res[1] + res[2] + res[3]);
  return 0;
}

File: bench/conc.rs
// Concurrency benchmark (Rust, std::thread) — the fork-join baseline: 4 workers
// each compute fib(N), the parent sums. `conc N T`: T<=1 runs them sequentially
// (the 1-thread baseline), else one std::thread per worker. Same workload as
// bench/conc.c and bench/conc.axi. Naive fib so the compute dominates.
use std::thread;

fn fib(n: i64) -> i64 {
    if n < 2 {
        n
    } else {
        fib(n - 1) + fib(n - 2)
    }
}

fn main() {
    let a: Vec<String> = std::env::args().collect();
    let n: i64 = a.get(1).and_then(|s| s.parse().ok()).unwrap_or(34);
    let t: usize = a.get(2).and_then(|s| s.parse().ok()).unwrap_or(4);
    // `black_box` so the optimizer cannot hoist/CSE the four `fib(n)` calls in the
    // sequential path (they must each really run, like the four session workers).
    let sum: i64 = if t <= 1 {
        (0..4).map(|_| fib(std::hint::black_box(n))).sum()
    } else {
        let hs: Vec<_> = (0..4)
            .map(|_| thread::spawn(move || fib(std::hint::black_box(n))))
            .collect();
        hs.into_iter().map(|h| h.join().unwrap()).sum()
    };
    println!("{sum}");
}
